src/main/java/de/tum/in/www1/artemis/service/CourseScoreCalculationService.java

```diff
@@ -198,20 +362,20 @@ Result getResultForParticipation(Participation participation, ZonedDateTime dueD
                 return ratedResults.get(0);
             }
 
-            // sorting in descending order to have the last result at the beginning
-            resultsList.sort(Comparator.comparing(Result::getCompletionDate).reversed());
+            Result latestResult = Collections.max(resultsList, Comparator.comparing(Result::getCompletionDate));
 
             long gracePeriodInSeconds = 10L;
-            if (dueDate == null || !dueDate.plusSeconds(gracePeriodInSeconds).isBefore(resultsList.get(0).getCompletionDate())) {
+            if (dueDate == null || (latestResult.getCompletionDate() != null && !dueDate.plusSeconds(gracePeriodInSeconds).isBefore(latestResult.getCompletionDate()))) {
```

Comment:
This is not correct.

Firstly, we should not duplicate the grace period logic here. Every result has a `rated` boolean, this should be used here.

If the latest result is not rated, there could be still another earlier result, which should be taken into consideration instead. The else-cause which sets the result to 0 points is wrong:

![grafik](https://user-images.githubusercontent.com/26540346/225681735-f30e2b02-62ab-4efd-9216-9788967dd4cf.png)

Reply:
- `dueDate` is null
This means that you are always allowed to submit something. This also means that the latstet result is always graded, it's not possible to create a result "after" this non-exisiting du date.

- compeltionDate is null
That's a good question, I'm not sure how and if this can happen. I'd simply refer to the `rated` boolean here.

- result.getCompletionDate() is after dueDate but before dueDate + 10 seconds
I'm not aware of such a grace period at other places. `setRatedIfNotExceeeded` uses the direct due date, I'm working on #6018 which will increase this period.

In any case, such a grace period should be consistently handled in the result (here we talk again about the rated boolean). Having such a special logic here would require it at quite a few other places dealing with results (e.g displaying it on the client)

Reply:
I got the `null` completionDate once on a test server for one participation which let the whole calculation fail though I have no idea how it got there in the first place, but it feels better to at least return a 0 Result in that case. 
I refactored the method to go through all the rated results until it finds one that has a completionDate before the due date. Should I just remove the grace period for now? It feels a little risky without knowing what its purpose was.